weather_dl/download_pipeline/fetcher.py

```diff
@@ -44,29 +44,122 @@ class Fetcher(beam.DoFn):
         client_name: The name of the download client to construct per each request.
         manifest: A manifest to keep track of the status of requests
         store: To manage where downloads are persisted.
+        optimise_download: Optimised the download.
     """
 
     client_name: str
     manifest: Manifest = NoOpManifest(Location('noop://in-memory'))
     store: t.Optional[Store] = None
+    optimise_download: bool = False
 
     def __post_init__(self):
         if self.store is None:
             self.store = FSStore()
 
+    def expand(self, element):
+        request = (
+            element
+            | 'EachPartition' >> beam.ParDo(EachPartition())
+        )
+
+        if self.optimise_download:
+            downloaded_data = (
+                request
+                | 'Fetch' >> beam.ParDo(FetchData(self.client_name, self.manifest, self.store))
+                | 'Download' >> beam.ParDo(DownloadData(self.client_name, self.manifest, self.store))
+            )
+        else:
+            downloaded_data = (
+                request
+                | 'Fetch+Download' >> beam.ParDo(Retrieve(self.client_name, self.manifest, self.store))
+            )
+        (
+            downloaded_data
+            | 'Upload' >> beam.ParDo(Upload(self.client_name, self.manifest, self.store))
+        )
+
+
+class EachPartition(beam.DoFn):
+    def process(self, element) -> t.Iterator[t.Tuple[t.Any, str]]:
+        # element: Tuple[Tuple[str, int], Iterator[Config]]
+        """Execute download requests one-by-one."""
+        (subsection, request_idx), partitions = element
+        worker_name = f'{subsection}.{request_idx}'
+        logger.info(f"[{worker_name}] Starting requests...")
+        for partition in partitions:
+            beam.metrics.Metrics.counter('Fetcher', subsection).inc()
+            yield (partition, worker_name)
+
+
+@dataclasses.dataclass
+class FetchData(beam.DoFn):
+    client_name: str
+    manifest: Manifest = NoOpManifest(Location('noop://in-memory'))
+    store: t.Optional[Store] = None
+
     @retry_with_exponential_backoff
-    def upload(self, src: io.FileIO, dest: str) -> None:
-        """Upload blob to cloud storage, with retries."""
-        with self.store.open(dest, 'wb') as dest_:
-            shutil.copyfileobj(src, dest_, WRITE_CHUNK_SIZE)
+    def fetch(self, client: Client, dataset: str, selection: t.Dict) -> t.Dict:
+        """Fetch data from download client, with retries."""
+        return client.fetch(dataset, selection)
+
+    def process(self, element) -> t.Iterator[t.Tuple[t.Dict, Config, str, str, str]]:
+        """Fetch data from client."""
+        config, worker_name = element
+
+        if not config:
+            return
+
+        if skip_partition(config, self.store):
+            return
+
+        client = CLIENTS[self.client_name](config)
+        target = prepare_target_name(config)
+
+        with self.manifest.transact(config.selection, target, config.user_id):
```

Comment:
A complication that I didn't anticipate until now: It would probably be best if we updated the manifest to distinguish between retrieved and downloaded. WDYT?

Reply:
Made necessary changes. Added a new class variable in DownloadStatus named stage which represents the current stage of the request.